File: Metaflow/ConcurrencyScopeGrain.fs
namespace Metaflow

open Metaflow
open Orleans
open FSharp.Control.Tasks



type ConcurrencyScopeGrain(clusterClient: IClusterClient) =
    inherit Grain()

    interface IConcurrencyScopeGrain with
        member this.Execute<'op, 'model, 'input>(call: FeatureCall<'input>) =
            task { return! Features.execute<'op, 'model, 'input> call clusterClient }

File: Metaflow.Orleans.Grains/ConcurrencyScopeGrain.cs
using System.Threading.Tasks;
using Microsoft.FSharp.Core;
using Orleans;

namespace Metaflow.Orleans
{
    public class ConcurrencyScopeGrain : Grain, IConcurrencyScopeGrain
    {
        private readonly IClusterClient _clusterClient;

        public ConcurrencyScopeGrain(IClusterClient clusterClient)
        {
            _clusterClient = clusterClient;
        }

        public Task<FSharpResult<Unit, FeatureFailure>> Execute<TOp, TModel>(
            FeatureCall<TModel> call)
        {
            return Workflows.executeFeature<TOp, TModel>(call, _clusterClient);
        }
    }
}

